diff -Naur awssdk-build.old/aws-cpp-sdk-core/source/utils/event/EventStreamDecoder.cpp awssdk-build/aws-cpp-sdk-core/source/utils/event/EventStreamDecoder.cpp
--- awssdk-build.old/aws-cpp-sdk-core/source/utils/event/EventStreamDecoder.cpp	2023-02-22 11:43:33.493965856 -0800
+++ awssdk-build/aws-cpp-sdk-core/source/utils/event/EventStreamDecoder.cpp	2023-02-22 10:30:53.561549772 -0800
@@ -72,9 +72,7 @@
                 assert(handler);
                 if (!handler)
                 {
-                    AWS_LOGSTREAM_ERROR(EVENT_STREAM_DECODER_CLASS_TAG, "Payload received, but decoder encountered internal errors before."
-                        "ErrorCode: " << EventStreamErrorsMapper::GetNameForError(handler->GetInternalError()) << ", "
-                        "ErrorMessage: " << handler->GetEventPayloadAsString());
+                    AWS_LOGSTREAM_ERROR(EVENT_STREAM_DECODER_CLASS_TAG, "Payload received, but handler is null.");
                     return;
                 }
                 handler->WriteMessageEventPayload(static_cast<unsigned char*>(payload->buffer), payload->len);
@@ -129,9 +127,7 @@
                 assert(handler);
                 if (!handler)
                 {
-                    AWS_LOGSTREAM_ERROR(EVENT_STREAM_DECODER_CLASS_TAG, "Payload received, but decoder encountered internal errors before."
-                        "ErrorCode: " << EventStreamErrorsMapper::GetNameForError(handler->GetInternalError()) << ", "
-                        "ErrorMessage: " << handler->GetEventPayloadAsString());
+                    AWS_LOGSTREAM_ERROR(EVENT_STREAM_DECODER_CLASS_TAG, "Header received, but handler is null.");
                     return;
                 }
 
